src/WebJobs.Script/Workers/WorkerConcurrencyManager.cs

```diff
@@ -0,0 +1,218 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Script.Workers.Rpc;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.WebJobs.Script.Workers
+{
+    internal class WorkerConcurrencyManager : IHostedService, IDisposable
+    {
+        private readonly TimeSpan _logStateInterval = TimeSpan.FromSeconds(60);
+        private readonly IOptions<WorkerConcurrencyOptions> _concurrencyOptions;
+        private readonly ILogger _logger;
+        private readonly IFunctionInvocationDispatcherFactory _functionInvocationDispatcherFactory;
+
+        private IFunctionInvocationDispatcher _functionInvocationDispatcher;
+        private System.Timers.Timer _timer;
+        private Stopwatch _addWorkerStopWatch = Stopwatch.StartNew();
+        private Stopwatch _logStateStopWatch = Stopwatch.StartNew();
+        private bool _disposed = false;
+
+        public WorkerConcurrencyManager(IFunctionInvocationDispatcherFactory functionInvocationDispatcherFactory,
+            IOptions<WorkerConcurrencyOptions> concurrencyOptions, ILoggerFactory loggerFactory)
+        {
+            _concurrencyOptions = concurrencyOptions ?? throw new ArgumentNullException(nameof(concurrencyOptions));
+            _functionInvocationDispatcherFactory = functionInvocationDispatcherFactory ?? throw new ArgumentNullException(nameof(functionInvocationDispatcherFactory));
+
+            _logger = loggerFactory?.CreateLogger<WorkerConcurrencyManager>();
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            if (_concurrencyOptions.Value.Enabled)
+            {
+                // Delay monitoring
+                await Task.Delay(_concurrencyOptions.Value.AdjustmentPeriod);
```

Comment:
This will delay host startup and impact coldstart. Can't do this delay here. Shouldn't be needed though, since you're already checking in the timer callback?

Reply:
good catch